### PR TITLE
OpenACC `parallel_for` and `parallel_reduce`

### PR DESCRIPTION
**Pull request by JBludau**
This PR advances the integration of the OpenACC backend into kokkos.

It introduces `parallel_for` and `parallel_reduce` for `RangePolicy` and bumps the incremental tests to level 5.
Pedro did the implementation work, I restructured, rebased and deleted includes some unnecessary code.

A lot of the implementation was based on `OpenMPTarget` so please think about if the design chosen there is something we want to stick with.

### CODE REVIEW

**Comment by JBludau** [comment]
@pedrovalerolara @seyonglee the branch is now rebased on the current develop now as requested

File: core/src/OpenACC/Kokkos_OpenACC_ParallelFor_Range.hpp
Comment (dalg24) [inline core/src/OpenACC/Kokkos_OpenACC_ParallelFor_Range.hpp:78]: Why "gang vector"?

File: core/src/OpenACC/Kokkos_OpenACC_ParallelFor_Range.hpp
Comment (JBludau) [inline core/src/OpenACC/Kokkos_OpenACC_ParallelFor_Range.hpp:71]: I added the `throw` here. Turns out, we should not throw when running a parallel for with size 0. Nevertheless, we should think about throwing for negative ranges, as this is something the user should be informed about in my opinion. Could be done at another point in core, as this should be independent of the backend.

File: core/src/OpenACC/Kokkos_OpenACC_ParallelReduce_Range.hpp
Comment (JBludau) [inline core/src/OpenACC/Kokkos_OpenACC_ParallelReduce_Range.hpp:109]: This could be collapsed into a single function returning the value which would be easier to read. Furthermore, this would reduce the number of times we get `begin`, `end`, etc.

File: core/src/OpenACC/Kokkos_OpenACC_DeepCopy.hpp
Comment (dalg24) [reply core/src/OpenACC/Kokkos_OpenACC_DeepCopy.hpp:153]: Does not compile w/o it?

File: core/src/OpenACC/Kokkos_OpenACC_ParallelFor_Range.hpp
Comment (seyonglee) [reply core/src/OpenACC/Kokkos_OpenACC_ParallelFor_Range.hpp:78]: OpenACC provides three levels of parallelism (gang/worker/vector), but most of existing GPUs (NVIDIA GPU and AMD GPU) provides only two levels of parallelism, and thus mapping OpenACC three level parallelism to the GPU's two level parallelism has been issues for a long time (and each OpenACC compiler (e.g., NVHPC and GCC) also changed their mapping strategies multiple times). For now, just using both gang and vector seems to be the best choice at least for both NVHPC and GCC.

Comment (seyonglee) [reply core/src/OpenACC/Kokkos_OpenACC_ParallelFor_Range.hpp:78]: `acc parallel loop …` may work well for NVHPC, but GCC seems to be less efficient in handling auto parallelization and thus may not work well.
A bigger issue is that an OpenACC compiler may not be able to figure out what kind of parallelisms are included in the functor, in which case, the compiler will not be able to find which parallelisms can be allowed to this loop. 
Therefore, it will be better to keep `loop gang vector` as it is.

**Comment by Rombur** [comment]
CI is failing
```
33: [ RUN      ] OPENACC.IncrTest_05_reduction
33: unknown file: Failure
33: C++ exception with description "Kokkos::Impl::ParallelFor< OpenACC > can not be executed with a range <= 0." thrown in the test body.
```

File: core/src/OpenACC/Kokkos_OpenACC_ParallelFor_Range.hpp
Comment (Rombur) [reply core/src/OpenACC/Kokkos_OpenACC_ParallelFor_Range.hpp:71]: Well looks like the CI hits this `throw`

Comment (masterleinad) [reply core/src/OpenACC/Kokkos_OpenACC_ParallelFor_Range.hpp:71]: I'm not quite sure if running with `end>begin` (and expecting to do nothing) is necessarily a bug.

Comment (pedrovalerolara) [reply core/src/OpenACC/Kokkos_OpenACC_ParallelFor_Range.hpp:71]: According to today's meeting, it looks like we will let this as it is  (the end <= begin is controlled by the back ends), right?

File: core/src/OpenACC/Kokkos_OpenACC_ParallelReduce_Range.hpp
Comment (pedrovalerolara) [reply core/src/OpenACC/Kokkos_OpenACC_ParallelReduce_Range.hpp:109]: Sorry Jakob, I need more details. What do you mean with "this could be collapsed"? All the "reduce" functions in one?

File: core/src/OpenACC/Kokkos_OpenACC_ParallelFor_Range.hpp
Comment (pedrovalerolara) [reply core/src/OpenACC/Kokkos_OpenACC_ParallelFor_Range.hpp:78]: It can look weird using the first (gang) and last (vector) OpenACC parallelism level together, but as Seyong commented, using gang and vector works better

File: core/src/OpenACC/Kokkos_OpenACC_ParallelReduce_Range.hpp
Comment (JBludau) [reply core/src/OpenACC/Kokkos_OpenACC_ParallelReduce_Range.hpp:109]: I meant, that the body of the `init` function could be put into the reduce function, thus only leaving a single function.

File: core/src/OpenACC/Kokkos_OpenACC_ParallelFor_Range.hpp
Comment (JBludau) [reply core/src/OpenACC/Kokkos_OpenACC_ParallelFor_Range.hpp:71]: > According to today's meeting, it looks like we will let this as it is (the end <= begin is controlled by the back ends), right?

We will just handle the case of `begin=end` in the backends and check for `end<begin` in `RangePolicy` (separate pr)

File: core/src/OpenACC/Kokkos_OpenACC_ParallelFor_Range.hpp
Comment (crtrott) [reply core/src/OpenACC/Kokkos_OpenACC_ParallelFor_Range.hpp:78]: do a ifdef based on compiler: for nvhpc use `acc parallel loop`, and for gcc `acc parallel loop gang vector`

Comment (crtrott) [reply core/src/OpenACC/Kokkos_OpenACC_ParallelFor_Range.hpp:78]: and then lets complain to the vendors. So generally I want to see at least a comment what we think we should ideally use, vs what we need to use because the compiler is dumb

Comment (seyonglee) [reply core/src/OpenACC/Kokkos_OpenACC_ParallelFor_Range.hpp]: `acc parallel loop gang vector` is the best choice for NVHPC too, and even the NVHPC compiler may have the same limit on handling functors; if the compiler cannot figure out available parallelisms in the functor, `acc parallel loop` will not work.

**Review by crtrott — APPROVED** [review]
Approving with the understanding that a whole bunch of reduction features are not yet supported.

File: core/src/OpenACC/Kokkos_OpenACC_ParallelReduce_Range.hpp
Comment (dalg24) [inline core/src/OpenACC/Kokkos_OpenACC_ParallelReduce_Range.hpp:63]: Note that we could go for a compile error if that is being instantiated (instead of runtime error).
I just refactored the initial approach and did not think it through.  Let me know if you want me to change now.

Comment (JBludau) [reply core/src/OpenACC/Kokkos_OpenACC_ParallelReduce_Range.hpp:63]: would prefer, as there is no necessity to wait on runtime

File: core/src/OpenACC/Kokkos_OpenACC_FunctorAdapter.hpp
Comment (JBludau) [inline core/src/OpenACC/Kokkos_OpenACC_FunctorAdapter.hpp:50]: Could be moved into a file that does not have `OpenACC` as there are no details depending on openACC and this could be used for other backends as well

**Review by JBludau — COMMENTED** [review]
I can't approve as author but looks good to me.

File: core/src/OpenACC/Kokkos_OpenACC_FunctorAdapter.hpp
Comment (dalg24) [reply core/src/OpenACC/Kokkos_OpenACC_FunctorAdapter.hpp:50]: My intent was to do that eventually (not in this PR)

File: core/src/OpenACC/Kokkos_OpenACC_ParallelReduce_Range.hpp
Comment (dalg24) [reply core/src/OpenACC/Kokkos_OpenACC_ParallelReduce_Range.hpp]: Done

File: core/src/OpenACC/Kokkos_OpenACC_ParallelReduce_Range.hpp
Comment (dalg24) [inline core/src/OpenACC/Kokkos_OpenACC_ParallelReduce_Range.hpp:161]: Oooops

File: core/src/OpenACC/Kokkos_OpenACC_FunctorAdapter.hpp
Comment (JBludau) [reply core/src/OpenACC/Kokkos_OpenACC_FunctorAdapter.hpp:50]: sure

File: core/src/OpenACC/Kokkos_OpenACC_ParallelReduce_Range.hpp
Comment (JBludau) [reply core/src/OpenACC/Kokkos_OpenACC_ParallelReduce_Range.hpp:161]: looks still okay though :-)

File: core/src/OpenACC/Kokkos_OpenACC_FunctorAdapter.hpp
Comment (masterleinad) [inline core/src/OpenACC/Kokkos_OpenACC_FunctorAdapter.hpp:78]:
```suggestion
template <class Functor, class Policy>
class FunctorAdapter {
  Functor m_functor;
  using WorkTag = typename Policy::work_tag;

 public:
  FunctorAdapter(Functor const &functor) : m_functor(functor) {}

  template <class... Args>
  KOKKOS_FUNCTION void operator()(Args &&... args) const {
    if constexpr (std::is_same_v<WorkTag>)
      m_functor(static_cast<Args &&>(args)...);
    else
      m_functor(WorkTag(), static_cast<Args &&>(args)...);
  }
};
```

File: core/src/OpenACC/Kokkos_OpenACC_ParallelReduce_Range.hpp
Comment (pedrovalerolara) [reply core/src/OpenACC/Kokkos_OpenACC_ParallelReduce_Range.hpp]: Sure! We can do that. Do you want me to do it? There some lines to change here, one per supported reduction type.

File: core/src/OpenACC/Kokkos_OpenACC_FunctorAdapter.hpp
Comment (dalg24) [reply core/src/OpenACC/Kokkos_OpenACC_FunctorAdapter.hpp:78]: Did you format?

Comment (masterleinad) [reply core/src/OpenACC/Kokkos_OpenACC_FunctorAdapter.hpp]: No, I didn't.

Comment (dalg24) [reply core/src/OpenACC/Kokkos_OpenACC_FunctorAdapter.hpp]: That's fine I had to edit anyway because the condition on the if constexpr was wrong.
I applied your suggestion though

**Review by seyonglee — APPROVED** [review]
It looks good to me too.

**Comment by dalg24** [comment]
```
4: [ RUN      ] openmptarget.int64_t_reduce_dynamic
4: CUDA error: an illegal memory access was encountered 
4: Libomptarget error: Call to targetDataEnd failed, abort target.
4: Libomptarget error: Failed to process data after launching the kernel.
4: Libomptarget error: Consult https://openmp.llvm.org/design/Runtimes.html for debugging options.
4: Kokkos_OpenMPTarget_Parallel.hpp:333:1: Libomptarget fatal error 1: failure of target construct while offloading is mandatory
 4/33 Test  #4: KokkosCore_UnitTest_OpenMPTarget .............Child aborted***Exception:  25.47 sec
```
and
```
7: [ RUN      ] cuda.debug_pin_um_to_host
7: Time CudaSpace: 0.006351 CudaUVMSpace_1: 0.006383 CudaUVMSpace_2: 0.015805 CudaPinnedHostSpace: 0.244889 CudaUVMSpace_Pinned: 0.018538
7: /var/jenkins/workspace/Kokkos/core/unit_test/cuda/TestCuda_DebugPinUVMSpace.cpp:128: Failure
7: Value of: passed
7:   Actual: false
7: Expected: true
7: [  FAILED  ] cuda.debug_pin_um_to_host (412 ms)
```
are clearly unrelated

**Pull request merged**
